### ISSUE TITLE
Tool calling issues with vllm

### ISSUE THREAD

**Issue by Navanit-git** [pydantic/pydantic-ai#224]
Hi, 
Is there a way where we can add the vLLM OpenAI compatibility support. 
[vllm OpenAI Support](https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html?ref=blog.mozilla.ai)

So that anyone can use any llm model calls.

**Comment by samuelcolvin** [comment]
Should work the same as ollama, see the code here - https://github.com/pydantic/pydantic-ai/issues/112#issuecomment-2524245937.

Happy to consider adding vllm as another custom model, but it would need more people to want it before we do the work.

**Comment by samuelcolvin** [comment]
See #239, that would mean we could add `VLLMModel`.

**Comment by daavoo** [comment]
> See #239, that would mean we could add `VLLMModel`.

Ola! I was testing pydantic.ai alongside vLLM and [llama.cpp server](https://github.com/ggerganov/llama.cpp/tree/master/examples/server), which I think both fulfill the requirements.

I have looked at the existing [Ollama code](https://github.com/pydantic/pydantic-ai/blob/main/pydantic_ai_slim/pydantic_ai/models/ollama.py) and I am not sure I understand the value of adding a new `VLLMModel`. 
Looks like there is no custom logic and just provides an ~arbitrary hardcoded list of model names. 

I got this same snippet working with both vLLM and llama.cpp server out of the box:



```bash
# For example with llama.cpp:
docker run -v ./models:/models -p 8080:8080 \
ghcr.io/ggerganov/llama.cpp:server -m /models/smollm2-360m-instruct-q8_0.gguf
```

```py
from pydantic_ai import Agent
from pydantic_ai.models.openai import OpenAIModel

model = OpenAIModel(
    "mymodel", 
    base_url="http://localhost:8080/v1", 
    api_key="foo"
)

agent = Agent(  
    model=model,
    system_prompt='Be concise, reply with one sentence.',  
)

result = agent.run_sync('Where does "hello world" come from?')  
print(result.data)
```



So, is it better to just send a small documentation patch? 

pd. I don't have a problem on contributing a new `VLLM/LLAMACPP` model myself.

**Comment by sadransh** [comment]
@daavoo This wouldnt work with tool calling and non str result type.  you can re-use the example from here: to double check https://github.com/pydantic/pydantic-ai/issues/398

**Comment by lys791227** [comment]
> @daavoo This wouldnt work with tool calling and non str result type. you can re-use the example from here: to double check #398

I also same.

**Comment by guang11644331** [comment]
from pydantic_ai import Agent
from pydantic_ai.models.openai import OpenAIModel


model = OpenAIModel(
    model_name="qwen25", 
    base_url="http://192.168.xx.xx:8090/v1", 
    api_key="1"
)

agent = Agent(  
    model=model,
)

result = agent.run_sync('hello')  
print(result.data)

##
it's work

**Comment by lys791227** [comment]
> from pydantic_ai import Agent from pydantic_ai.models.openai import OpenAIModel
> 
> model = OpenAIModel( model_name="qwen25", base_url="http://192.168.xx.xx:8090/v1", api_key="1" )
> 
> agent = Agent( model=model, )
> 
> result = agent.run_sync('hello') print(result.data)
> 
> it's work
Do you call tool try?should not !

**Comment by iabgcb** [comment]
Does not work with tool calling, excluding which there is no reason for enterprises/retail hosting their own quantized
models to use this framework. so please VLLM compatibility ASAP

**Comment by maziyarpanahi** [comment]
Obviously, it works for some basic use that doesn't require `tool` selection, but even a simple code like this fails with vLLM:

```
from pydantic import BaseModel

from pydantic_ai import Agent
from pydantic_ai.models.openai import OpenAIModel

model = OpenAIModel(
   model_name="qwen25",
   base_url="http://192.168.xx.xx:8090/v1",
   api_key="1"
)

class CityLocation(BaseModel):
    city: str
    country: str

agent = Agent(model=model, result_type=CityLocation)

result = agent.run_sync('Where were the olympics held in 2012?')
print(result.data)
print(result.usage())
```
 
error (`tool_choice must either be a named tool, "auto", or "none"`)

```
openai.BadRequestError: Error code: 400 - {'object': 'error', 'message': '[{\'type\': \'value_error\', \'loc\': (\'body\',), \'msg\': \'Value error, `tool_choice` must either be a named tool, "auto", or "none".\', \'input\': {\'messages\': [{\'role\': \'user\', \'content\': \'Where were the olympics held in 2012?\'}], \'model\': \'arcee-train/arcee-virtuoso-maz-10b-v2\', \'n\': 1, \'parallel_tool_calls\': True, \'stream\': False, \'tool_choice\': \'required\', \'tools\': [{\'type\': \'function\', \'function\': {\'name\': \'final_result\', \'description\': \'The final response which ends this conversation\', \'parameters\': {\'properties\': {\'city\': {\'title\': \'City\', \'type\': \'string\'}, \'country\': {\'title\': \'Country\', \'type\': \'string\'}}, \'required\': [\'city\', \'country\'], \'title\': \'CityLocation\', \'type\': \'object\'}}}]}, \'ctx\': {\'error\': ValueError(\'`tool_choice` must either be a named tool, "auto", or "none".\')}}]', 'type': 'BadRequestError', 'param': None, 'code': 400}
```

**Comment by iabgcb** [comment]
exactly the same error i receive. It simply does not work with vllm.

**Comment by nihilimbo** [comment]
I can confirm the sample code from @maziyarpanahi [comment](https://github.com/pydantic/pydantic-ai/issues/224#issuecomment-2596153497) fails on my dev box with a "tool_choice must either be a named tool, auto, or none" error when run against a vLLM (v0.6.6.post1) instance of meta-llama/Llama-3.3-70B-Instruct.

However, same code works fine against an Ollama instance (on same machine) of llama3.1:8b.

**Comment by myyang19770915** [comment]
Yes , I use vllm model,  and get the same tool calling problems.
please add this work in this project !
thanks

**Comment by TheCodingLand** [comment]
I think it's related to the way vllm handles tool calling.

https://docs.vllm.ai/en/latest/features/tool_calling.html

Perhaps only a documentation item is required if we find a compatible option for the vllm server backend.

**Comment by daavoo** [comment]
> I think it's related to the way vllm handles tool calling.
> 
> https://docs.vllm.ai/en/latest/features/tool_calling.html
> 
> Perhaps only a documentation item is required if we find a compatible option for the vllm server backend.

Did you guys tried launching vllm with the arguments shown in that link?

**Comment by maziyarpanahi** [comment]
I have tried it. The error is about `'tool_choice': 'required'`. What you see in vLLM is to activate tool_choice `auto`. You can add these 2 to enable this in vLLM:

```
    --enable-auto-tool-choice \
    --tool-call-parser hermes \
```

But it doesn't help since `pydantic-ai` is asking for `required` when it comes to tool_choice:

```
`tool_choice` must either be a named tool, "auto", or "none"
```

**Comment by sydney-runkle** [comment]
We could probably fix the tool choice issue by further expanding `ModelSettings`.

**Comment by iabgcb** [comment]
would love to help.

**Comment by maziyarpanahi** [comment]
FYI: this is a PR to follow: https://github.com/pydantic/pydantic-ai/pull/825

**Comment by sydney-runkle** [comment]
Update: I think this should be fixed by #825. We aren't planning on adding a custom model for vllm, as we've removed the `OllamaModel` in favor of recommending usage with `OpenAIModel` as the two are compatible.

See more docs here: https://ai.pydantic.dev/models/#ollama

**Comment by BeautyyuYanli** [comment]
Here is my workaround: https://gist.github.com/BeautyyuYanli/9e66513e665ebaf3a4658e61fb9c04ef

**Comment by tomliptrot** [comment]
I think the relevant vLLM documentation is this: https://docs.vllm.ai/en/latest/features/structured_outputs.html The JSON schema should be getting passed to vLLM via the `extra_body` paramater, rather than via tools (I may be wrong):

```python
from pydantic import BaseModel
from enum import Enum

class CarType(str, Enum):
    sedan = "sedan"
    suv = "SUV"
    truck = "Truck"
    coupe = "Coupe"


class CarDescription(BaseModel):
    brand: str
    model: str
    car_type: CarType


json_schema = CarDescription.model_json_schema()

completion = client.chat.completions.create(
    model="Qwen/Qwen2.5-3B-Instruct",
    messages=[
        {
            "role": "user",
            "content": "Generate a JSON with the brand, model and car_type of the most iconic car from the 90's",
        }
    ],
    extra_body={"guided_json": json_schema},
)
print(completion.choices[0].message.content)
```

**Comment by nihilimbo** [comment]
Any progress on this? vLLM is the best local inference engine for many people and we'd really love to be able to use Pydantic-AI (remote APIs are not an option in some use cases). The [PR](https://github.com/pydantic/pydantic-ai/pull/825) seems to be pretty stuck.

**Comment by arty-hlr** [comment]
This https://github.com/vllm-project/vllm/pull/13483 was just merged...

**Comment by ItzAmirreza** [comment]
Sorry, I'm a bit confused. 

> This [vllm-project/vllm#13483](https://github.com/vllm-project/vllm/pull/13483) was just merged...

Do I have to do something or if I update my vllm it will work automatically? I'm still getting
'tool_choice` must either be a named tool, "auto", or "none"'

**Comment by arty-hlr** [comment]
@ItzAmirreza You updated vllm and it's still not working?

**Comment by ItzAmirreza** [comment]
> [@ItzAmirreza](https://github.com/ItzAmirreza) You updated vllm and it's still not working?

I'm not sure how to update, I'm using the vllm/vllm-openai:latest on docker hub which is updated 10 days ago. Do you have any idea how can I test the new one?

**Comment by suresh-now** [comment]
@ItzAmirreza , I am still having issue. I am using docker.io/vllm/vllm-openai:v0.8.3 with - --enable-auto-tool-choice
- --tool-call-parser
- llama3_json, I still have issue, Below is by code

 `
api_key = os.getenv('API_KEY')
base_url = os.getenv('BASE_URL')
llm_model_name = os.getenv('LLM_MODEL_NAME')
base_url = "url-of-model"

model = OpenAIModel(
    llm_model_name,
    provider=OpenAIProvider(
        base_url=base_url, api_key=api_key
    ),
)

class CityLocation(BaseModel):
    city: str
    country: str

async def run_agent():
    agent = Agent(model=model, result_type=CityLocation)
    result = await agent.run('What is the capital of France?',infer_name="true")
    print(result.data)
    print(result.usage())



if __name__ == "__main__":
    asyncio.run(run_agent())
`

I get this error ```status_code: 400, model_name: meta-llama/Llama-3.2-3B-Instruct, body: {'object': 'error', 'message': 'The provided JSON schema contains features not supported by xgrammar.', 'type': 'BadRequestError', 'param': None, 'code': 400}```, seems like we need to extra_body parameter to model - https://docs.vllm.ai/en/latest/features/structured_outputs.html , any way to fix this?

**Comment by suresh-now** [comment]
I was able to fix the above issue by deploying vllm using - --guided_decoding_backend "outlines". This will fix structured output format.

Also pydantic doesnt work well with, 
https://docs.vllm.ai/en/stable/features/tool_calling.html#models-with-pythonic-tool-calls-pythonic. Any ways to make it work?

**Comment by Kludex** [comment]
I think the issue described in the description of this issue was solved, so I'll be closing this.

If that's not true, please provide an MRE, and I'm happy to reopen.
